modules/test/portlet-container-test/test/integration/com/liferay/portlet/container/test/BaseTestCase.java

```diff
@@ -0,0 +1,243 @@
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.portlet.container.test;
+
+import com.liferay.portal.kernel.test.rule.DeleteAfterTestRun;
+import com.liferay.portal.kernel.test.util.GroupTestUtil;
+import com.liferay.portal.kernel.test.util.TestPropsValues;
+import com.liferay.portal.kernel.util.WebKeys;
+import com.liferay.portal.model.Company;
+import com.liferay.portal.model.Group;
+import com.liferay.portal.model.Layout;
+import com.liferay.portal.model.Portlet;
+import com.liferay.portal.security.permission.PermissionChecker;
+import com.liferay.portal.security.permission.PermissionCheckerFactoryUtil;
+import com.liferay.portal.security.permission.PermissionThreadLocal;
+import com.liferay.portal.service.CompanyLocalServiceUtil;
+import com.liferay.portal.theme.ThemeDisplay;
+import com.liferay.portal.theme.ThemeDisplayFactory;
+import com.liferay.portal.util.test.LayoutTestUtil;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+import java.net.HttpURLConnection;
+import java.net.URL;
+
+import java.util.Arrays;
+import java.util.Dictionary;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.util.tracker.ServiceTracker;
+
+import org.springframework.mock.web.MockHttpServletRequest;
+
+/**
+ * @author Raymond Augé
+ */
+public class BaseTestCase {
+
+	@Before
+	public void setUp() throws Exception {
+		group = GroupTestUtil.addGroup();
+
+		layout = LayoutTestUtil.addLayout(group);
+
+		PermissionChecker permissionChecker =
+			PermissionCheckerFactoryUtil.create(TestPropsValues.getUser());
+
+		PermissionThreadLocal.setPermissionChecker(permissionChecker);
+	}
+
+	@After
+	public void tearDown() throws Exception {
+		PermissionThreadLocal.setPermissionChecker(null);
+
+		if (serviceTracker != null) {
+			serviceTracker.close();
+		}
+
+		for (ServiceRegistration<?> serviceRegistration :
+				serviceRegistrations) {
+
+			serviceRegistration.unregister();
+		}
+
+		serviceRegistrations.clear();
+	}
+
+	protected String drain(InputStream stream) throws IOException {
+		if (stream == null) {
+			return "";
+		}
+
+		byte[] bytes = new byte[100];
+		StringBuffer buffer = new StringBuffer(500);
+		int length = 0;
+
+		while ((length = stream.read(bytes)) != -1) {
+			String chunk = new String(bytes, 0, length);
```

Comment:
Ok, that's fine, I am just trying to say, test code is also code, we should treat it in the same way as production code. After all, when things fails, no matter it is production or test code caused, someone has to work on it. Writing a better test code is just helping future yourself and everyone else.